readme
1.It needs libmemcached 0.40, patched with fail_over.patch, empty_string.patch, buffer_request.patch
2.If you use the python whose version is < 2.5, you should modify the Py_ssize_t definition.


ChangeLog
0.40 Mon Mar 15 13:36:00 CST 2011
    * update to libmemcached-0.40
	* support large item (>=1M), auto splitting and merging
	* set_multi, prepend_multi, delete_multi
	* nonblock by default
	* fail-over patch for libmemcached-0.40
	* empty string patch for libmemcached-0.40
	* configable item compression and decompression when needed
	* use marshal prior to cPickle when serializing.

0.23 Mon Aug 25 16:57:59 CST 2008
	* Support the libmemcached 0.23 version
	* server rehash when one server is down
	* weighted ketama
	* cas support and fix incr and decr bug

0.14 Sun Mar 16 22:03:10 CST 2008
	* Frist douban version


test.py
import cmemcached
import unittest
import pickle

class TestCmemcached(unittest.TestCase):

    def setUp(self, with_cas=0):
        self.mc = cmemcached.Client(["127.0.0.1:11211"], behaviors={'support_cas': with_cas})

    def testSetAndGet(self):
        self.mc.set("num12345", 12345)
        self.assertEqual(self.mc.get("num12345"), 12345)
        self.mc.set("str12345", "12345")
        self.assertEqual(self.mc.get("str12345"), "12345")

    def testDelete(self):
        self.mc.set("str12345", "12345")
        #delete return True on success, otherwise False
        ret = self.mc.delete("str12345")
        self.assertEqual(self.mc.get("str12345"), None)
        self.assertEqual(ret, True)

        ret = self.mc.delete("hello world")
        self.assertEqual(ret, False)

    def testGetMulti(self):
        self.mc.set("a", "valueA")
        self.mc.set("b", "valueB")
        self.mc.set("c", "valueC")
        result = self.mc.get_multi(["a", "b", "c", "", "hello world"])
        self.assertEqual(result, {'a':'valueA', 'b':'valueB', 'c':'valueC'})

    def testBigGetMulti(self):
        count = 10
        keys = ['key%d' % i for i in xrange(count)]
        pairs = zip(keys, ['value%d' % i for i in xrange(count)])
        for key, value in pairs:
            self.mc.set(key, value)
        result = self.mc.get_multi(keys)
        assert result == dict(pairs)

    def testGetList(self):
        self.mc.delete("")
        self.mc.delete("hello world")
        self.mc.set("a", "valueA")
        self.mc.set("b", "valueB")
        self.mc.set("c", "valueC")
        result= self.mc.get_list(["a", "b", "c", "", "hello world"])
        self.assertEqual(result, ['valueA', 'valueB', 'valueC', None, None])

    def testAppend(self):
        self.mc.delete("a")
        self.mc.set("a", "I ")
        ret = self.mc.append("a", "Do")
        result = self.mc.get("a")
        self.assertEqual(result, "I Do")

    def testPrepend(self):
        self.mc.delete("a")
        self.mc.set("a", "Do")
        ret = self.mc.prepend("a", "I ")
        result = self.mc.get("a")
        self.assertEqual(result, "I Do")

    def testIncr(self):
        self.mc.set("incr", 1)
        ret = self.mc.incr("incr", 1)
        self.assertEqual(ret, 2)
        ret = self.mc.incr("incr", 2)
        self.assertEqual(ret, 4)

    def testDecr(self):
        self.mc.set("decr", 10)
        ret = self.mc.decr("decr", 1)
        self.assertEqual(ret, 9)
        ret = self.mc.decr("decr", 2)
        self.assertEqual(ret, 7)

    def testBook(self):
        self.mc.set("bool", True)
        value = self.mc.get("bool")
        self.assertEqual(value, True)
        self.mc.set("bool_", False)
        value = self.mc.get("bool_")
        self.assertEqual(value, False)

    def testBehaviors(self):
        mc2 = cmemcached.Client(["127.0.0.1:11211"])
        self.assertEqual(mc2.behaviors['support_cas'], 0)
        mc2 = cmemcached.Client(["127.0.0.1:11211"],
                behaviors={'support_cas': 1})
        self.assertEqual(mc2.behaviors['support_cas'], 1)
        mc2 = cmemcached.Client(["127.0.0.1:11211"],
                behaviors={'support_cas': 0})
        self.assertEqual(mc2.behaviors['support_cas'], 0)


class TestCmemcachedCas(TestCmemcached):
    """
    Run the tests again, but with cas support on the default connection.
    """

    def setUp(self):
        super(TestCmemcachedCas, self).setUp(with_cas=1)
        self.mc_cas = self.mc

    def testCas(self):
        self.mc_cas.set('miredo', 'alamos')
        (res, cas), = self.mc_cas.get_list(['miredo'], with_cas=True)
        r2, c2 = self.mc_cas.gets('miredo')
        self.assertEqual((res, cas), (r2, c2))
        self.assertEqual(res, 'alamos')
        self.assertFalse(self.mc_cas.cas('miredo', 'monobot', 0))
        self.assertEqual(self.mc_cas.get('miredo'), 'alamos')
        self.assertTrue(self.mc_cas.cas('miredo', 'oulipo', cas))
        self.assertEqual(self.mc_cas.get('miredo'), 'oulipo')

if __name__ == '__main__':
    unittest.main()



t/benchmark.py
#!/usr/bin/env python

import time
import random
import string
import sys
import traceback


global total_time

def run_test(func, name):
    sys.stdout.write(name + ': ')
    sys.stdout.flush()
    start_time = time.time()
    try:
        func()
    except:
        print "failed or not supported"
        global options
        if options.verbose:
            traceback.print_exc()
    else:
        end_time = time.time() 
        global total_time
        total_time += end_time - start_time
        print "%f seconds" % (end_time - start_time)


class BigObject(object):
    def __init__(self, letter='1', size=10000):
        self.object = letter * size

    def __eq__(self, other):
        return self.object == other.object


class Benchmark(object):
    def __init__(self, module, options):
        self.module = module
        self.options = options
        self.init_server()
        self.test_set()
        self.test_set_get()
        self.test_random_get()
        self.test_set_same()
        self.test_set_big_object()
        self.test_set_get_big_object()
        self.test_set_big_string()
        self.test_set_get_big_string()
        self.test_get()
        self.test_get_big_object()
        self.test_get_multi()
        self.test_get_list()

    def init_server(self):
        #self.mc = self.module.Client([self.options.server_address])
        self.mc = self.module.Client(["127.0.0.1:11211","127.0.0.1:11212", "127.0.0.1:11213", "127.0.0.1:11214", "127.0.0.1:11215"])
        self.mc.set('bench_key', "E" * 50)

        num_tests = self.options.num_tests
        self.keys = ['key%d' % i for i in xrange(num_tests)]
        self.values = ['value%d' % i for i in xrange(num_tests)]
        import random
        self.random_keys = ['key%d' % random.randint(0, num_tests) for i in xrange(num_tests * 3)]

    def test_set(self):
        set_ = self.mc.set
        pairs = zip(self.keys, self.values)

        def test():
            for key, value in pairs:
                set_(key, value)
        def test_loop():
            for i in range(10):
                for key, value in pairs:
                    set_(key, value)
        run_test(test, 'test_set')

        for key, value in pairs:
            self.mc.delete(key)

    def test_set_get(self):
        set_ = self.mc.set
        get_ = self.mc.get
        pairs = zip(self.keys, self.values)

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value
        run_test(test, 'test_set_get')

        #for key, value in pairs:
        #    self.mc.delete(key)

    def test_random_get(self):
        get_ = self.mc.get
        set_ = self.mc.set

        value = "chenyin"

        def test():
            index = 0
            for key in self.random_keys:
                result = get_(key)
                index += 1
                if(index % 5 == 0):
                    set_(key, value)
        run_test(test, 'test_random_get')

    def test_set_same(self):
        set_ = self.mc.set

        def test():
            for i in xrange(self.options.num_tests):
                set_('key', 'value')
        def test_loop():
            for i in range(10):
                for i in xrange(self.options.num_tests):
                    set_('key', 'value')
        run_test(test, 'test_set_same')

        self.mc.delete('key')

    def test_set_big_object(self):
        set_ = self.mc.set
        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, BigObject()) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)

        run_test(test, 'test_set_big_object (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)

    def test_set_get_big_object(self):
        set_ = self.mc.set
        get_ = self.mc.get
        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, BigObject()) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value

        run_test(test, 'test_set_get_big_object (100 objects)')

        #for key, value in pairs:
        #    self.mc.delete(key)

    def test_set_get_big_string(self):
        set_ = self.mc.set
        get_ = self.mc.get

        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, 'x' * 10000) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value
        run_test(test, 'test_set_get_big_string (100 objects)')


    def test_set_big_string(self):
        set_ = self.mc.set

        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, 'x' * 10000) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
        run_test(test, 'test_set_big_string (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)


    def test_get(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        get = self.mc.get

        def test():
            for key, value in pairs:
                result = get(key)
                assert result == value
        run_test(test, 'test_get')

        for key, value in pairs:
            self.mc.delete(key)
        
    def test_get_big_object(self):
        pairs = [('bkey%d' % i, BigObject('x')) for i in xrange(100)]
        for key, value in pairs:
            self.mc.set(key, value)

        get = self.mc.get
        expected_values = [BigObject('x') for i in xrange(100)]
        
        def test():
            for i in xrange(100):
                result = get('bkey%d' % i)
                assert result == expected_values[i]
        run_test(test, 'test_get_big_object (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)

    def test_get_multi(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        keys = self.keys
        expected_result = dict(pairs)

        def test():
            result = self.mc.get_multi(keys)
            assert result == expected_result
        run_test(test, 'test_get_multi')

        for key, value in pairs:
            self.mc.delete(key)

    def test_get_list(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        keys = self.keys
        expected_result = self.values
        
        def test():
            result = self.mc.get_list(keys)
            assert result == expected_result
        run_test(test, 'test_get_list')
        
        for key in self.keys:
            self.mc.delete(key)


def main():
    from optparse import OptionParser
    parser = OptionParser()
    parser.add_option('-a', '--server-address', dest='server_address',
            default='127.0.0.1:11211',
            help="address:port of memcached [default: 127.0.0.1:11211]")
    parser.add_option('-n', '--num-tests', dest='num_tests', type='int',
            default=1000,
            help="repeat counts of each test [default: 1000]")
    parser.add_option('-v', '--verbose', dest='verbose',
            action='store_true', default=False,
            help="show traceback infomation if a test fails")
    global options
    options, args = parser.parse_args()


    options.verbose=1
    print "verbose", options.verbose

    global total_time
    total_time = 0

    print "Benchmarking cmemcached..."
    import cmemcached
    Benchmark(cmemcached, options)


if __name__ == '__main__':
    main()
    global total_time
    print "total_time is %f" % total_time


t/pickle_perf.py
from cPickle import dumps, loads
from timeit import Timer

num = 300
class BigObject(object):
    def __init__(self, letter='1', size=10000):
        self.object = letter * size

    def __eq__(self, other):
        return self.object == other.object

sillything = []

for i in (1000, 10000, 100000, 200000):
    value = [BigObject(str(j%10), i) for j in xrange(num)]
    sillything.append(value)


value_t = None

def test_pickle():
    for tset in sillything:
        for t in tset:
            value = dumps(t, -1)
            value_t = loads(value)
            #assert value_t == t

from time import time
t=time()
for i in xrange(10):
    test_pickle()
print time()-t
#t=Timer(setup='from __main__ import test_pickle', stmt='test_pickle()')
#print t.timeit()



split_mc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "split_mc.h"

int split_mc_set(struct memcached_st *mc, char *key, size_t key_len, void *val,
		size_t bytes, time_t expire, uint32_t flags)
{
	char chunk_key[250];
	int i, retval;
	size_t chunk_bytes;

	/* assert bytes > CHUNK_SIZE */
    if (key_len > 200 || bytes > CHUNK_SIZE * 10) return -1; 

	for (i=0; bytes>0; i++) {
		retval = snprintf(chunk_key, 250, "~%.*s/%d", (int)key_len, key, i);
		if (retval < 0 || retval >= 250) {
			return -1;
		}
		chunk_bytes = bytes > CHUNK_SIZE ? CHUNK_SIZE : bytes;
		retval = memcached_set(mc, chunk_key, retval, val, chunk_bytes,
				expire, flags);
		if (retval != 0 && retval != 31) {
			return retval;
		}
		val += CHUNK_SIZE;
		bytes -= chunk_bytes;
	}

	sprintf(chunk_key, "%d", i);  /* re-use chunk_key as value buffer */
	retval = memcached_set(mc, key, key_len, chunk_key, strlen(chunk_key)+1,
			expire, flags|FLAG_CHUNKED);

	return retval;
}

/* It's up to the caller to free returned *val !!! */
char* split_mc_get(struct memcached_st *mc, char *key, size_t key_len,
		int nchunks, size_t *bytes)
{
	int i;
	char *c_val, *r, *v;
	memcached_return rc;
	uint32_t flags;
	size_t length;
	char chunk_key[250];

	/* assert res && res.val && (res.flags & FLAG_CHUNKED) */
    if (nchunks > 10 || key_len > 200)
        return NULL;

	r = (char*)malloc(sizeof(char) * CHUNK_SIZE * nchunks);
	if (!r) return NULL;

	for (i=0, v=r; i<nchunks; i++) {
		snprintf(chunk_key, 250, "~%.*s/%d", (int)key_len, key, i);
		c_val = memcached_get(mc, chunk_key, strlen(chunk_key),
				&length, &flags, &rc);	
		if(rc != MEMCACHED_SUCCESS || !c_val || length > CHUNK_SIZE) {
            if (c_val) free(c_val);
			goto error;
		}

		memcpy(v, c_val, length);
        free(c_val);
		v += length;
	}

    *bytes = v - r;
	return r;

error:
	memcached_delete(mc, key, key_len, 0);
	free(r);
	return NULL;
}


cmemcached.py
import sys
from zlib import compress, decompress, error as zlib_error
from cmemcached_imp import *
import cmemcached_imp 

_FLAG_PICKLE = 1<<0
_FLAG_INTEGER = 1<<1
_FLAG_LONG = 1<<2
_FLAG_BOOL = 1<<3
_FLAG_COMPRESS = 1<<4
_FLAG_MARSHAL = 1<<5

VERSION="0.40"

def prepare(val, comp_threshold):
    val, flag = cmemcached_imp.prepare(val)
    if comp_threshold > 0 and val and len(val) > comp_threshold:
        val = compress(val)
        flag |= _FLAG_COMPRESS
    return val, flag

def restore(val, flag):
    if val is None:
        return val

    if flag & _FLAG_COMPRESS:
        try:
            val = decompress(val)
        except zlib_error:
            return None
        flag &= ~_FLAG_COMPRESS

    return cmemcached_imp.restore(val, flag)

class Client(cmemcached_imp.Client):
    "a wraper around cmemcached_imp"

    def __init__(self, servers, do_split=1, comp_threshold=0, *a, **kw):
        cmemcached_imp.Client.__init__(self)
        self.add_server(servers)
        self.do_split = do_split
        self.comp_threshold = comp_threshold
        
        self.set_behavior(BEHAVIOR_NO_BLOCK, 1) # nonblock
        self.set_behavior(BEHAVIOR_TCP_NODELAY, 1) # nonblock
        self.set_behavior(BEHAVIOR_TCP_KEEPALIVE, 1)
        self.set_behavior(BEHAVIOR_CACHE_LOOKUPS, 1)
        self.set_behavior(BEHAVIOR_KETAMA, 1)
        #self.set_behavior(BEHAVIOR_BUFFER_REQUESTS, 0) # no request buffer
        
    def set(self, key, val, time=0, compress=True):
        comp = compress and self.comp_threshold or 0
        val, flag = prepare(val, comp)
        if val is not None:
            return self.set_raw(key, val, time, flag)
        else:
            print >>sys.stderr, '[cmemcached]', 'serialize %s failed' % key

    def set_multi(self, values, time=0, compress=True):
        comp = compress and self.comp_threshold or 0
        raw_values = dict((k, prepare(v, comp)) for k,v in values.iteritems())
        return self.set_multi_raw(raw_values, time)

    def get(self, key):
        val, flag = cmemcached_imp.Client.get_raw(self, key)
        return restore(val, flag)

    def get_multi(self, keys):
        result = cmemcached_imp.Client.get_multi_raw(self, keys)
        return dict((k, restore(v, flag))
                    for k, (v, flag) in result.iteritems())

    def get_list(self, keys):
        result = self.get_multi(keys)
        return [result.get(key) for key in keys]


benchmark.py
#!/usr/bin/env python

import time
import random
import string
import sys


global total_time

def run_test(func, name):
    sys.stdout.write(name + ': ')
    sys.stdout.flush()
    start_time = time.time()
    try:
        func()
    except:
        print "failed or not supported"
        global options
        if options.verbose:
            import traceback; traceback.print_exc()
    else:
        end_time = time.time() 
        global total_time
        total_time += end_time - start_time
        print "%f seconds" % (end_time - start_time)


class BigObject(object):
    def __init__(self, letter='1', size=10000):
        self.object = letter * size

    def __eq__(self, other):
        return self.object == other.object


class Benchmark(object):
    def __init__(self, module, options):
        self.module = module
        self.options = options
        self.init_server()
        self.test_set()
        self.test_set_get()
        self.test_random_get()
        self.test_set_same()
        self.test_set_big_object()
        self.test_set_get_big_object()
        self.test_set_big_string()
        self.test_set_get_big_string()
        self.test_get()
        self.test_get_big_object()
        self.test_get_multi()
        self.test_get_list()

    def init_server(self):
        self.mc = self.module.Client([self.options.server_address])
        self.mc.set('bench_key', "E" * 50)

        num_tests = self.options.num_tests
        self.keys = ['key%d' % i for i in xrange(num_tests)]
        self.values = ['value%d' % i for i in xrange(num_tests)]
        import random
        self.random_keys = ['key%d' % random.randint(0, num_tests) for i in xrange(num_tests * 3)]

    def test_set(self):
        set_ = self.mc.set
        pairs = zip(self.keys, self.values)

        def test():
            for key, value in pairs:
                set_(key, value)
        def test_loop():
            for i in range(10):
                for key, value in pairs:
                    set_(key, value)
        run_test(test, 'test_set')

        for key, value in pairs:
            self.mc.delete(key)

    def test_set_get(self):
        set_ = self.mc.set
        get_ = self.mc.get
        pairs = zip(self.keys, self.values)

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value
        run_test(test, 'test_set_get')

        #for key, value in pairs:
        #   self.mc.delete(key)

    def test_random_get(self):
        get_ = self.mc.get
        set_ = self.mc.set

        value = "chenyin"

        def test():
            index = 0
            for key in self.random_keys:
                result = get_(key)
                index += 1
                if(index % 5 == 0):
                    set_(key, value)
        run_test(test, 'test_random_get')

    def test_set_same(self):
        set_ = self.mc.set

        def test():
            for i in xrange(self.options.num_tests):
                set_('key', 'value')
        def test_loop():
            for i in range(10):
                for i in xrange(self.options.num_tests):
                    set_('key', 'value')
        run_test(test, 'test_set_same')

        self.mc.delete('key')

    def test_set_big_object(self):
        set_ = self.mc.set
        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, BigObject()) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)

        run_test(test, 'test_set_big_object (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)

    def test_set_get_big_object(self):
        set_ = self.mc.set
        get_ = self.mc.get
        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, BigObject()) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value

        run_test(test, 'test_set_get_big_object (100 objects)')

        #for key, value in pairs:
        #   self.mc.delete(key)

    def test_set_get_big_string(self):
        set_ = self.mc.set
        get_ = self.mc.get

        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, 'x' * 10000) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
                result = get_(key)
                assert result == value
        run_test(test, 'test_set_get_big_string (100 objects)')


    def test_set_big_string(self):
        set_ = self.mc.set

        # libmemcached is slow to store large object, so limit the
        # number of objects here to make tests not stall.
        pairs = [('key%d' % i, 'x' * 10000) for i in xrange(100)]

        def test():
            for key, value in pairs:
                set_(key, value)
        run_test(test, 'test_set_big_string (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)


    def test_get(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        get = self.mc.get

        def test():
            for key, value in pairs:
                result = get(key)
                assert result == value
        run_test(test, 'test_get')

        for key, value in pairs:
            self.mc.delete(key)
        
    def test_get_big_object(self):
        pairs = [('bkey%d' % i, BigObject('x')) for i in xrange(100)]
        for key, value in pairs:
            self.mc.set(key, value)

        get = self.mc.get
        expected_values = [BigObject('x') for i in xrange(100)]
        
        def test():
            for i in xrange(100):
                result = get('bkey%d' % i)
                assert result == expected_values[i]
        run_test(test, 'test_get_big_object (100 objects)')

        for key, value in pairs:
            self.mc.delete(key)

    def test_get_multi(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        keys = self.keys
        expected_result = dict(pairs)

        def test():
            result = self.mc.get_multi(keys)
            assert result == expected_result
        run_test(test, 'test_get_multi')

        for key, value in pairs:
            self.mc.delete(key)

    def test_get_list(self):
        pairs = zip(self.keys, self.values)
        for key, value in pairs:
            self.mc.set(key, value)

        keys = self.keys
        expected_result = self.values
        
        def test():
            result = self.mc.get_list(keys)
            assert result == expected_result
        run_test(test, 'test_get_list')
        
        for key in self.keys:
            self.mc.delete(key)


def main():
    from optparse import OptionParser
    parser = OptionParser()
    parser.add_option('-a', '--server-address', dest='server_address',
            default='127.0.0.1:11211',
            help="address:port of memcached [default: 127.0.0.1:11211]")
    parser.add_option('-n', '--num-tests', dest='num_tests', type='int',
            default=1000,
            help="repeat counts of each test [default: 1000]")
    parser.add_option('-v', '--verbose', dest='verbose',
            action='store_true', default=False,
            help="show traceback infomation if a test fails")
    global options
    options, args = parser.parse_args()

    global total_time
    total_time = 0

    print "Benchmarking cmemcached..."
    import cmemcached
    Benchmark(cmemcached, options)


if __name__ == '__main__':
    main()
    global total_time
    print "total_time is %f" % total_time
